File: app/src/main/java/agrawal/bhanu/jetpack/reddit/ui/RedditPostViewModel.java
package agrawal.bhanu.jetpack.reddit.ui;

import android.app.Application;
import androidx.arch.core.util.Function;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;
import androidx.annotation.NonNull;

import java.util.concurrent.Executor;

import javax.inject.Inject;

import agrawal.bhanu.jetpack.reddit.data.ItemKeyedPostDataSource;
import agrawal.bhanu.jetpack.MyApp;
import agrawal.bhanu.jetpack.reddit.data.PostDataSourceFactory;
import agrawal.bhanu.jetpack.network.model.NetworkState;
import agrawal.bhanu.jetpack.reddit.model.Post;
import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class RedditPostViewModel extends ViewModel {

    PostDataSourceFactory postDataSourceFactory;
    private LiveData<PagedList<Post>> postList;
    private LiveData<NetworkState> networkState;
    private LiveData<NetworkState> initloading;




    @Inject
    public RedditPostViewModel(PostDataSourceFactory postDataSourceFactory) {
        super();
        this.postDataSourceFactory = postDataSourceFactory;
    }

    public LiveData<PagedList<Post>> getPostList() {

        if(postList == null){
            PagedList.Config pagedListConfig =
                    (new PagedList.Config.Builder()).setEnablePlaceholders(false)
                            .setInitialLoadSizeHint(20)
                            .setPageSize(20).build();

            postList = (new LivePagedListBuilder(postDataSourceFactory, pagedListConfig))
                    .build();
        }
        return postList;
    }

    public LiveData<NetworkState> getNetworkState() {

        networkState = Transformations.switchMap(postDataSourceFactory.getMutableLiveData(), new Function<ItemKeyedPostDataSource, LiveData<NetworkState>>() {
            @Override
            public LiveData<NetworkState> apply(ItemKeyedPostDataSource input) {
                return input.getNetworkState();
            }
        });

        return networkState;
    }

    public LiveData<NetworkState> getInitloading() {
       initloading = Transformations.switchMap(postDataSourceFactory.getMutableLiveData(), new Function<ItemKeyedPostDataSource, LiveData<NetworkState>>() {
           @Override
           public LiveData<NetworkState> apply(ItemKeyedPostDataSource input) {
               return input.getInitloading();
           }
       });
       return initloading;
    }


    public void onRefresh() {
        postDataSourceFactory.getMutableLiveData().getValue().invalidate();
    }

    public void retry(){
        postDataSourceFactory.getMutableLiveData().getValue().retry();
    }
}

File: app/src/main/java/agrawal/bhanu/jetpack/reddit/ui/RedditPostViewModel.kt
package agrawal.bhanu.jetpack.reddit.ui

import agrawal.bhanu.jetpack.network.model.NetworkState
import agrawal.bhanu.jetpack.reddit.data.ItemKeyedPostDataSource
import agrawal.bhanu.jetpack.reddit.data.PostDataSourceFactory
import agrawal.bhanu.jetpack.reddit.model.Post
import androidx.arch.core.util.Function
import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

@HiltViewModel
class RedditPostViewModel @Inject constructor(
        var postDataSourceFactory: PostDataSourceFactory
) : ViewModel() {

    var postList: LiveData<PagedList<Post>?>? = null
        get() {
            if (field == null) {
                val pagedListConfig = PagedList.Config.Builder().setEnablePlaceholders(false)
                        .setInitialLoadSizeHint(20)
                        .setPageSize(20).build()
                field = LivePagedListBuilder(postDataSourceFactory, pagedListConfig)
                        .build()
            }
            return field
        }
        private set

    val networkState: LiveData<NetworkState?> = Transformations.switchMap(
            postDataSourceFactory.mutableLiveData
        ){
            it.networkState as LiveData<NetworkState?>
        }
    val initloading: LiveData<NetworkState?> = Transformations.switchMap(
            postDataSourceFactory.mutableLiveData
    ){
        it.initloading as LiveData<NetworkState?>
    }


    fun onRefresh() {
        postDataSourceFactory.mutableLiveData.value!!.invalidate()
    }

    fun retry() {
        postDataSourceFactory.mutableLiveData.value!!.retry()
    }
}